rfcs/000-third-party-software-installation.md

```diff
@@ -0,0 +1,109 @@
+
+Additionally, we propose the use of `~/.v` and the relevant subfolders (like `~/.v/modules`, `~/.v/bin`) to store the source code, built binaries, etc.
+
+# Motivation
+
+Currently, V only allows downloading the source code for V modules in the `VMODULES` directory through the `v install` command.
+This is good for making modules globally available. There is currently no streamlined way of _installing_ or _fetching_ or _obtaining_ third-party software written in V.
+
+# Guide-level explanation
+
+Take [_klonol_](https://github.com/hungrybluedev/klonol) for example.
+It is a command-line tool written in V that helps to maintain a local copy of all git repositories on GitHub and Gitea.
+If you want to install it, you have to clone the source yourself, run `v build.vsh` and then add the `bin` folder to `PATH`.
+A nice thing to have is a `v fetch <repo>` or `v obtain <repo>` command (or something else, please comment) that will clone the repo and build it for you.
+
+Only once, like when V is install, for example, the user needs to add `~/.v/bin` to `PATH`.
+This will make all third-party applications installed using `v fetch <repo>` available from anywhere.
+
+# Reference-level explanation
+
+It will piggyback off of `v install`.
+The act of installing the software is done through the following additional steps:
+
+1. Running `v build.vsh` or some other specified build script in the root of the repo.
+2. This will create an executable, along with all other
+   relevant files (like templates, etc) in the `bin` directory of the repo. Ideally, the resources will be embedded in the executable to allow for only one executable to be enough.
+3. The executable will be added to `~/.v/bin` for the user to call from anywhere.
+   TODO.
+   An additional **suggestion** is to have the `~/.vmodules` directory be replaced in favour of `~/.v/modules`.
+
+If there is another script (other than `build.vsh`) that generates the executable
+
+A sample `build.vsh` script is:
+
+```text
+println('Removing old artifacts...')
+rmdir_all('bin') or {}
+println('Done removing "bin" directory.')
+
+println('\nCreating new output directory "bin"...')
+mkdir('bin') ?
+println('Done creating "bin" directory.')
+
+println('\nChecking if everything is formatted correctly...')
+execute_or_panic('v fmt -verify .')
+println('Done checking formatting.')
+
+println('\nCompiling and building executable...')
+execute_or_panic('v -prod . -o bin/klonol')
+println('Done compiling and placing executable in "bin".')
+```
+
+Taken from [klonol](https://github.com/hungrybluedev/klonol/blob/6e36e5676542aae40e0124de13055643484bb015/build.vsh).
+
+Users will have to add `~/.v/bin` to their `PATH` once.
+
+We also make use of the `bin` directory to keep the other files in the root of the repository out of the search path.
+
+# Drawbacks
+
+- Security concerns. Perhaps users should take some extra steps to install software from third-party sources.
```

Comment:
This is one of the criticism also fo cargo that pull randomly dependencies, However, here we need to split the problem in different areas.

1. pull modules from Git, in this case the user need to trust the git repo
2. pull modules from a package repository, in this case all the publisher should be verified

Reply:
Limitations in the build.vsh could also work. Some yaml instead of a vsh. Through that the developer is limited in making damage to the system.

Reply:
We probably need to implement a YAML parser first then. Or warp an existing parser written in C. Alternatively, we can use TOML. But the spec should allow for a limited subset of actions like @danieldaeschle mentioned. One such (incomplete) list could be [this](https://github.com/vlang/rfcs/pull/22/#discussion_r833168027)

Reply:
>We probably need to implement a YAML parser first then. Or warp an existing parser written in C.

Better have one in pure V, but as the first start, we can include one written in C.

>Some yaml instead of a vsh. Through that the developer is limited in making damage to the system.

IMO a manifest file it is the way to go, the language can be in any type of data model language, but we need some things to define some rules

Reply:
I am against *any* configuration format like YAML, that uses significant whitespace.

Reply:
TOML is nice in part because unlike YAML, it does not mandate significant indentation/whitespace.

Reply:
> > We probably need to implement a YAML parser first then. Or warp an existing parser written in C.

No we do not. YAML is a retarded format. People should stop advocating its usage.
V's main repo already includes a good TOML parser, just do `import toml`, and you are ready to parse stuff.

Reply:
>No we do not. YAML is a retarded format. People should stop advocating its usage.
V's main repo already includes a good TOML parser, just do import toml, and you are ready to parse stuff.

I have no opinion on the type of data model we can use, for sure TOML is better also from my point of view. The more general point is that we need a manifest like a system, from my point of view it resolve a lot of problems